File: subasta-backend/models/Subasta.js
const mongoose = require('mongoose');

// Definir el esquema para cada Puesto dentro de la subasta
const PuestoSchema = mongoose.Schema({
    numero: {
        type: Number,
        required: true,
        min: 1,
        max: 100 // Limite de 100 puestos
    },
    ocupadoPor: { // ID del usuario que ocupa este puesto
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Usuario', // Asumiendo que tienes un modelo de Usuario
        default: null
    },
    montoPuja: { // La puja específica para este puesto
        type: Number,
        default: 0
    },
    fechaOcupacion: {
        type: Date,
        default: Date.now
    }
});

const SubastaSchema = mongoose.Schema({
    titulo: {
        type: String,
        required: true,
        trim: true
    },
    descripcion: {
        type: String,
        required: true
    },
    precioInicial: {
        type: Number,
        required: true,
        min: 0
    },
    // precioActual y ultimaPuja ya no son directamente aplicables a la subasta general
    // Ahora, cada puesto tendrá su propio monto de puja.
    // Podrías mantener precioActual si quieres que represente la puja más alta en general.
    precioActual: { // Se actualizará con la puja más alta entre todos los puestos
        type: Number,
        default: function() { return this.precioInicial; }
    },
    fechaInicio: {
        type: Date,
        default: Date.now
    },
    fechaFin: {
        type: Date,
        required: true
    },
    estado: {
        type: String,
        enum: ['activa', 'finalizada', 'cancelada'],
        default: 'activa'
    },
    // Ganador ya no es un solo ID, será un objeto con puesto y puja ganadora
    // ganador: { // Este campo se determinará al finalizar
    //     type: mongoose.Schema.Types.ObjectId,
    //     ref: 'Usuario',
    //     default: null
    // },
    // ultimaPuja ya no aplica de la misma manera
    // ultimaPuja: {
    //     type: mongoose.Schema.Types.ObjectId,
    //     ref: 'Puja',
    //     default: null
    // },
    imagenUrl: {
        type: String,
        default: null
    },
    // ¡¡¡NUEVO CAMPO CRÍTICO PARA LOS PUESTOS!!!
    puestos: {
        type: [PuestoSchema], // Array de objetos Puesto
        default: () => Array.from({ length: 100 }, (_, i) => ({
            numero: i + 1,
            ocupadoPor: null,
            montoPuja: 0,
            fechaOcupacion: null
        }))
    },
    // Campos para el ganador final (se llenan al finalizar)
    puestoGanador: {
        type: Number,
        default: null
    },
    pujaGanadora: {
        type: Number,
        default: null
    },
    ganadorId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Usuario',
        default: null
    }
});

module.exports = mongoose.model('Subasta', SubastaSchema);

File: frontend/app/src/main/java/com/example/subastaapp/model/Subasta.kt
package com.example.subastaapp.model

import com.google.gson.annotations.SerializedName
import java.util.Date

/**
 * Representa un puesto dentro de una subasta.
 */
data class Puesto(
    val numero: Int, // Número del puesto.
    val ocupadoPor: String? = null, // ID del ocupante.
    val montoPuja: Double, // Monto de la puja.
    val fechaOcupacion: Date? = null // Fecha de ocupación.
)

/**
 * Representa un usuario simplificado.
 */
data class UsuarioSimple(
    @SerializedName("_id")
    val id: String, // ID del usuario.
    val nombre: String // Nombre del usuario.
)

/**
 * Representa una subasta con todos sus detalles.
 */
data class Subasta(
    @SerializedName("_id")
    val id: String, // ID de la subasta.
    val titulo: String, // Título de la subasta.
    val descripcion: String, // Descripción.
    val precioInicial: Double, // Precio de inicio.
    val precioActual: Double, // Precio actual.
    val fechaInicio: Date, // Fecha de inicio.
    val fechaFin: Date, // Fecha de fin.
    val estado: String, // Estado.
    val imagenUrl: String? = null, // URL de la imagen.
    val puestos: List<Puesto>, // Lista de puestos.
    val puestoGanador: Int? = null, // Puesto ganador.
    val pujaGanadora: Double? = null, // Puja ganadora.
    val ganadorId: String? = null // ID del ganador.
)

/**
 * Datos para crear una nueva subasta.
 */
data class SubastaCreationRequest(
    val titulo: String, // Título.
    val descripcion: String, // Descripción.
    val precioInicial: Double, // Precio inicial.
    val fechaFin: Date, // Fecha de fin.
    val imagenUrl: String? = null // URL de la imagen (opcional).
)

/**
 * Solicitud para ocupar un puesto en una subasta.
 */
data class OcuparPuestoRequest(
    val puestoNumero: Int, // Número de puesto.
    val montoPuja: Double, // Monto de la puja.
    val pujadorId: String // ID del pujador.
)
